docs: update creating source plugins guide & add example repo (#22943)
gatsbyjs/gatsby: docs/docs/creating-a-source-plugin.md

### Caching data between runs

Some operations like fetching data from an endpoint can be performance heavy or time intensive, in order to improve the experience of developing with your source plugin, you can leverage the Gatsby cache to store data between runs of `gatsby develop` or `gatsby build`.

You access the `cache` in Gatsby Node APIs and use the `set` and `get` functions to store and retrive data as JSON objects:

```javascript:title=source-plugin/gatsby-node.js
exports.onPostBuild = async ({ cache }) => {
  await cache.set(`key`, `value`)
  const cachedValue = await cache.get(`key`)
  console.log(cachedValue) // logs `value`
}
```

> The above example shows a contrived example for the `cache`, but it can used in more sophisticated cases to reduce the time it takes to run your plugin. For example, by caching a timestamp, you can use it to fetch solely the data that has been updated from the data source:

Author: Edited your suggestion to add in a note about it being 'data since the last time data was fetched from the source'

  )
  // ...
}

exports.onPostBuild = async ({ cache }) => {
  // set a timestamp at the end of the build
  await cache.set(`timestamp`, Date.now())
}
```

> In addition to the cache, plugin's can save metadata to the [internal Redux store](/docs/data-storage-redux/) with `setPluginStatus`

This can reduce the time it takes repeated data fetching operations to run if you are pulling in large amounts of data for your plugin. Existing plugins like [`gatsby-source-contentful`](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-contentful/src/gatsby-node.js) generate a token that is sent with each request to only return new data.

You can read more about the cache API, other types of plugins that leverage the cache, and example open source plugins that use the cache in the [Build Caching](/docs/build-caching) guide.

### Adding relationships between nodes

Gatsby source plugins not only create nodes, they also create relationships between nodes that are exposed to GraphQL queries.

There are two types of node relationships in Gatsby: (1) foreign-key based and (1) transformations (parent-child).

#### Option 1: foreign-key relationships

> An example of a foreign-key relationship would be a Post that has an Author.

Author: Just updated this to clarify!

> In this relationship, each object is a distinct entity that exists whether or not the other does, with independent schemas, and field(s) on each entity that reference the other entity -- in this case the Post would have an Author, and the Author might have Posts. The API of a service that allows complex object modelling, for example a CMS, will often allow users to add relationships between entities and expose them through the API.

Author: I broke this down and added some graphql code fields to show what the objects would look like, how they are independent and then linked together which I think can help illustrate this better.

  })
}
```

For a stricter GraphQL schema, you can specify the exact field and value to link nodes with using schema customization APIs.

```javascript:title=source-plugin/gatsby-node.js
exports.sourceNodes = ({ actions, createContentDigest }) => {
  const { createNode } = actions
  createNode({
    // Data for the Post node
    // highlight-start
    author: {
      name: `Jay Gatsby`,
    },
    // highlight-end
    // Required fields
    id: `a-node-id`,
    parent: null
    children: [],
    internal: {
      type: `post`,
      contentDigest: createContentDigest(fieldData),
    }
  })
}

exports.createSchemaCustomization = ({ actions }) => {
  const { createTypes } = actions
  createTypes(`
    type Post implements Node {
      id: ID!
      # create a relationship between Post and the File nodes for optimized images
      author: Author @link(from: "author.name" by: "name") // highlight-line
      # ... other fields
    }`)
}
```

##### Creating the reverse relationship

It's often convenient for querying to add to the schema backwards references. For example, you might want to query the Author of a Post but you might also want to query all the posts an author has written.

> If you want to call this field on `Author` `posts` using the inference method, you would create a field called `posts___NODE` to hold the relationship to Posts. The value of this field should be an array of Post IDs.

Reviewer: Posts vs. `posts` here is a bit confusing. Can we be consistent?

Author: You're right this is confusing. I tried to update it to use the capital with backticks when it's referring to the GraphQL node type and to a lowercased posts when it's used to refer to posts more colloquially. How does it look now?

Here's an example from the [WordPress source plugin](https://github.com/gatsbyjs/gatsby/blob/1fb19f9ad16618acdac7eda33d295d8ceba7f393/packages/gatsby-source-wordpress/src/normalize.js#L178-L189).

> With schema customization, you would add the `@link` directive to your Author type and store the Post IDs on the Author nodes when they were created on a field used when types are created:

Reviewer: ```suggestion
With schema customization, you would add the `@link` directive to your Author type. Additionally, you can store the Post IDs on the Author nodes on a field when they are created.
```
Not sure I rephrased the last piece right. Open to other ideas.

```javascript:title=source-plugin/gatsby-node.js
exports.createSchemaCustomization = ({ actions }) => {
  const { createTypes } = actions
  createTypes(`
    type Post implements Node {
      id: ID!
      # create a relationship between Post and the File nodes for optimized images
      author: Author @link(from: "author.name" by: "name") // highlight-line
      # ... other fields
    }
    
    type Author implements Node {
      name: String!
      post: Post @link // highlight-line
    }`)
}
```

> You can read more about connecting foreign key fields with schema customization in the guide on [Customizing the GraphQL Schema](/docs/schema-customization/#foreign-key-fields).

Reviewer: ```suggestion
You can read more about connecting foreign key fields with schema customization in the guide on [customizing the GraphQL schema](/docs/schema-customization/#foreign-key-fields).
```

#### Option 2: transformation relationships

> Transformation relationships are used when a node is _completely_ derived from another node. An example that is common in source plugins is for transforming File nodes from remote sources, like for images from remote sources. You can read about this use case in the section below on [sourcing images from remote locations](/docs/creating-a-source-plugin/#sourcing-images-from-remote-locations).

Reviewer: ```suggestion
When a node is _completely_ derived from another node you'll want to use a transformation relationship. An example that is common in source plugins is for transforming File nodes from remote sources, e.g. images. You can read about this use case in the section below on [sourcing images from remote locations](/docs/creating-a-source-plugin/#sourcing-images-from-remote-locations).
```

> You can find more information about transformation relationships in the [Creating a Transformer Plugin guide](/docs/creating-a-transformer-plugin/#creating-the-transformer-relationship).

Reviewer: ```suggestion
You can find more information about transformation relationships in the [creating a transformer plugin guide](/docs/creating-a-transformer-plugin/#creating-the-transformer-relationship).
```

#### Union types

When creating fields linking to an array of nodes, if the array of IDs are all of the same type, the relationship field that is created will be of this type. If the linked nodes are of different types; the field will turn into a union type of all types that are linked. See the [GraphQL documentation on how to query union types](https://graphql.org/learn/schema/#union-types).

### Working with data received from remote sources

#### Setting media and MIME types

Each node created by the filesystem source plugin includes the raw content of the file and its _media type_.

> [A **media type**](https://en.wikipedia.org/wiki/Media_type) (also **MIME type** and **content type**) is an official way to identify the format of files/content that is transmitted on the internet, e.g. over HTTP or through email. You might be familiar with other media types such as `application/javascript`, `application/pdf`, `audio/mpeg`, `text/html`, `text/plain`, `image/jpeg`, etc.

Reviewer: ```suggestion
[A **media type**](https://en.wikipedia.org/wiki/Media_type) (also **MIME type** and **content type**) is an official way to identify the format of files/content that are transmitted via the internet, e.g. over HTTP or through email. You might be familiar with other media types such as `application/javascript`, `audio/mpeg`, `text/html`, etc.
```
three examples should suffice

> Each source plugin is responsible for setting the media type for the nodes they create. This way, source and transformer plugins can work together easily.

Reviewer: ```suggestion
Each source plugin is responsible for setting the media type for the nodes it creates. This way, source and transformer plugins can work together easily.
```

> This is not a required field -- if it's not provided, Gatsby will [infer](/docs/glossary#inference) the type from data that is sent -- but it's the way for source plugins to indicate to transformers that there is "raw" data that can still be further processed. It also allows plugins to remain small and focused. Source plugins don't have to have opinions on how to transform their data: they can set the `mediaType` and push that responsibility to transformer plugins, instead.

Reviewer: ```suggestion
This is not a required field -- if it's not provided, Gatsby will [infer](/docs/glossary#inference) the type from data that is sent -- but it's how source plugins indicate to transformers that there is "raw" data the transformer can further process. It also allows plugins to remain small and focused. Source plugins don't have to have opinions on how to transform their data: they can set the `mediaType` and push that responsibility to transformer plugins instead.
```

Reviewer: We might want to take the last two sentences and move them to the following paragraph. The transition is a bit abrupt.

For example, it's common for services to allow you to add content in Markdown format. If you pull that Markdown into Gatsby and create a new node, what then? How would a user of your source plugin convert that Markdown into HTML they can use in their site? You would create a node for the Markdown content and set its `mediaType` as `text/markdown` and the various Gatsby Markdown transformer plugins would see your node and transform it into HTML.

This loose coupling between the data source and the transformer plugins allow Gatsby site builders to assemble complex data transformation pipelines with little work on their (and your (the source plugin author)) part.

#### Sourcing and optimizing images from remote locations

> A common use case from transforming data from a remote source in source plugins is pulling images from a remote location and optimizing them for use with [Gatsby Image](/packages/gatsby-image/). An API may return a url for an image on a CDN, which could be further optimized by Gatsby at build time.

Reviewer: ```suggestion
A common use case for source plugins is pulling images from a remote location and optimizing them for use with [Gatsby Image](/packages/gatsby-image/). An API may return a URL for an image on a CDN, which could be further optimized by Gatsby at build time.
```

> This can be achieved through a number of steps with the use of several plugins:

Reviewer: ```suggestion
This can be achieved by the following steps:
```

```javascript:title=source-plugin/gatsby-node.js
const { createRemoteFileNode } = require(`gatsby-source-filesystem`)

exports.onCreateNode = async ({
  actions: { createNode },
  getCache,
  createNodeId,
  node,
}) => {
  // because onCreateNode is called for all nodes, verify that you are only running this code on nodes created by your plugin
  if (node.internal.type === `your-source-node-type`) {
    // create a FileNode in Gatsby that gatsby-transformer-sharp will create optimized images for
    const fileNode = await createRemoteFileNode({
      // the url of the remote image to generate a node for
      url: node.imgUrl,
      getCache,
      createNode,
      createNodeId,
      parentNodeId: node.id,
    })

    // highlight-start
    if (fileNode) {
      // with schemaCustomization: add a field `remoteImage` to your source plugin's node from the File node
      node.remoteImage = fileNode.id

      // OR with inference: link your source plugin's node to the File node without schemaCustomization like this, but creates a less sturdy schema
      node.remoteImage___NODE = fileNode.id
    }
    // highlight-end
  }
}
```

> Attaching `fileNode.id` to `remoteImage___NODE` will rely on Gatsby's [inference](/docs/glossary/#inference) of the GraphQL schema to create a new field `remoteImage` as a relationship between the nodes. This will be done automatically. For a sturdier schema, you can relate them using [`schemaCustomization` APIs](/docs/node-apis/#createSchemaCustomization) by adding the `fileNode.id` to a field that you reference when you `createTypes`:

Reviewer: ```suggestion
Attaching `fileNode.id` to `remoteImage___NODE` will rely on Gatsby's [inference](/docs/glossary/#inference) of the GraphQL schema to create a new field `remoteImage` as a relationship between the nodes. This is done automatically. For a sturdier schema, you can relate them using [`schemaCustomization` APIs](/docs/node-apis/#createSchemaCustomization) by adding the `fileNode.id` to a field that you reference when you `createTypes`:
```

exports.createSchemaCustomization = ({ actions }) => {
  const { createTypes } = actions
  createTypes(`
    type YourSourceType implements Node {
      id: ID!
      # create a relationship between YourSourceType and the File nodes for optimized images
      remoteImage: File @link // highlight-line
    }`)
}
```

4. Verify that `gatsby-plugin-sharp` and `gatsby-transformer-sharp` are included in the site that is using the plugin:

```javascript:title=gatsby-config.js
module.exports = {
  plugins: [
    // loads the source-plugin
    `your-source-plugin`,
    // required to generate optimized images
    `gatsby-plugin-sharp`,
    `gatsby-transformer-sharp`,
  ],
}
```

Then, the sharp plugins will automatically transform the File nodes created by `createRemoteFileNode` in `your-source-plugin` (which have supported image extensions like .jpg or .png). You can then query for the `remoteImage` field on your source type:

```graphql
query {
  yourSourceType {
    id
    remoteImage {
      childImageSharp {
        # fluid or fixed fields for optimzed images
      }
    }
  }
}
```

### Improve plugin developer experience by enabling faster sync

> One challenge when developing locally is that a developer might make modifications in a remote data source, like a CMS, and then want to see how it looks in the local environment. Typically they will have to restart the `gatsby develop` server to see changes. In order to improve the development experience of using a plugin, you can reduce the time it takes to sync between Gatsby and the data source by enabling faster synchronization of data changes from the source with the plugin. There are two approaches for doing this:

Reviewer: ```suggestion
One challenge when developing locally is that a developer might make modifications in a remote data source, like a CMS, and then want to see how it looks in the local environment. Typically they will have to restart the `gatsby develop` server to see changes. In order to improve the development experience of using a plugin, you can reduce the time it takes to sync between Gatsby and the data source by enabling faster synchronization of data changes. There are two approaches for doing this:
```

  if (pluginOptions.preview) {
    const subscription = await subscription(SUBSCRIPTION_TO_WEBSOCKET)
    subscription.subscribe(({ data: newData }) => {
      newData.forEach(newDatum => {
        switch (newDatum.status) {
          case "deleted":
            deleteNode({
              node: getNode(createNodeId(`YourSourceType-${newDatum.uuid}`)),
            })
            break
          case "created":
          case "updated":
          default:
            // created and updated can be handled by the same code path
            // the post's id is presumed to stay constant (or can be inferred)
            createNode(processDatum(newDatum))
            break
        )
      }
    })
  }
  // highlight-end

  const data = await client.query(QUERY_TO_API)

  // Process data and create nodes.using a custom processDatum function
  data.forEach(datum => createNode(processDatum(datum)))

  // You're done, return.
  return
}
```

_Note: This is pseudo code to illustrate the logic and concept of how these plugins function, you can see an example in the [creating source plugins](https://github.com/gatsbyjs/gatsby/tree/master/examples/creating-source-plugins) repository._

> Because the code in sourceNodes will be reinvoked when changes in the data source occur, a few steps need to be taken to make sure that Gatsby is tracking the existing nodes that have already been sourced along with the new data. A first step is ensuring that the existing nodes created are not garbage collected which is done by "touching" the nodes with the [`touchNode` action](/docs/actions/#touchNode).

Reviewer: ```suggestion
Because the code in `sourceNodes` is reinvoked when changes in the data source occur, a few steps need to be taken to ensure that Gatsby is tracking the existing nodes as well as the new data. A first step is ensuring that the existing nodes created are not garbage collected which is done by "touching" the nodes with the [`touchNode` action](/docs/actions/#touchNode).
```

> Then the new data needs to be pulled in via a live update like a websocket (in the example above with a subscription). The new data needs to have some information attached that dictates whether the data was created, updated, or deleted so that when it is processed a new node can be created/updated (with `createNode`) or deleted (with `deleteNode`). In the example above that information is coming from `newDatum.status`.

Reviewer: ```suggestion
Then the new data needs to be pulled in via a live update like a websocket (in the example above with a subscription). The new data needs to have some information attached that dictates whether the data was created, updated, or deleted; that way, when it is processed, a new node can be created/updated (with `createNode`) or deleted (with `deleteNode`). In the example above that information is coming from `newDatum.status`.
```

## Additional resources

- Working example repository on [creating source plugins](https://github.com/gatsbyjs/gatsby/tree/master/examples/creating-source-plugins) with the features in this guide implemented
- Tutorial on [Creating a Pixabay Image Source Plugin](/tutorial/pixabay-source-plugin-tutorial/)

> - [`gatsby-node-helpers`](https://github.com/angeloashmore/gatsby-node-helpers), a community-made NPM package with helper functions to generate Node objects with required fields like IDs and the `contentDigest` MD5 hash.

Reviewer: ```suggestion
- [`gatsby-node-helpers`](https://github.com/angeloashmore/gatsby-node-helpers), a community-made npm package with helper functions to generate Node objects with required fields like IDs and the `contentDigest` MD5 hash.
```

  const { createTypes } = actions
  createTypes(`
    type Post implements Node {
      id: ID!
      # create a relationship between Post and the File nodes for optimized images
      author: Author @link(from: "author.name" by: "name") // highlight-line
      # ... other fields
    }
    
    type Author implements Node {
      name: String!
      post: Post @link // highlight-line
    }`)
}
```

You can read more about connecting foreign key fields with schema customization in the guide on [customizing the GraphQL schema](/docs/schema-customization/#foreign-key-fields).

#### Option 2: transformation relationships

When a node is _completely_ derived from another node you'll want to use a transformation relationship. An example that is common in source plugins is for transforming File nodes from remote sources, e.g. images. You can read about this use case in the section below on [sourcing images from remote locations](/docs/creating-a-source-plugin/#sourcing-images-from-remote-locations).

You can find more information about transformation relationships in the [creating a transformer plugin guide](/docs/creating-a-transformer-plugin/#creating-the-transformer-relationship).

> #### Union types

Reviewer: Are union types another option? Additional? A bit confusing in the header hierarchy.

Author: They are additional, but I get your point that it's confusing yeah. I updated the text under it to explain it a little more, but I think it's a less common thing so didn't want to put it up in the section on adding relationships when it felt a little more advanced. 

It does correspond to both types of relationships though.
